### PR TITLE
feat(cli): run test modules in parallel

### PR DESCRIPTION
**Pull request by caspervonb**
This runs test modules in parallel when the number of concurrent jobs passed via `--jobs=<N>` is greater than 1.
The default is to run with a limit of 1 to retain backwards compatibility, as not all test suites in the wild are concurrency safe.

Closes https://github.com/denoland/deno/issues/6559
Closes https://github.com/denoland/deno/issues/9284

### CODE REVIEW

File: cli/main.rs
Comment (caspervonb) [inline cli/main.rs:3]: To self, remember to re-enable and lint.

File: cli/ops/test_dispatcher.rs
Comment (bartlomieju) [inline cli/ops/test_dispatcher.rs:39]: Could we get away without an op if module execution returned a `v8` value?

File: cli/program_state.rs
Comment (bartlomieju) [inline cli/program_state.rs:142]: Can this function could be reused for other purposes besides testing?

File: cli/test_dispatcher.rs
Comment (bartlomieju) [inline cli/test_dispatcher.rs:1]: Move to `cli/tools/test_runner.rs`

File: runtime/js/40_testing.js
Comment (bartlomieju) [inline runtime/js/40_testing.js:42]: This comment is useful

File: cli/program_state.rs
Comment (caspervonb) [reply cli/program_state.rs:142]: Should be usable for `cache_subcommand` as-well.

File: cli/ops/test_dispatcher.rs
Comment (caspervonb) [reply cli/ops/test_dispatcher.rs:39]: Yeah we could expose `tests` on Deno.internal and execute the tests from Rust instead (effectively moving the loop in runTests into Rust).

Would still use a channel like now but the sender would just be a local to `run_test`.

File: cli/tests/integration_tests.rs
Comment (caspervonb) [inline cli/tests/integration_tests.rs:2416]: Blocker

**Comment by caspervonb** [comment]
Had to run this a couple of times to get a decent sample.
We're failing with not found on reads, which is odd 🤔

**Comment by caspervonb** [comment]
Some quick and dirty benchmarks with `time` over `deno test` on deno_std.
Setting the number of jobs higher than the number of cores available more or less gives the same result within a reasonable margin of error so I cut it at 8.

```
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   40 bits physical, 48 bits virtual
CPU(s):                          6
On-line CPU(s) list:             0-5
Thread(s) per core:              1
Core(s) per socket:              6
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           61
Model name:                      Intel Core Processor (Broadwell, no TSX, IBRS)
Stepping:                        2
CPU MHz:                         2399.996
BogoMIPS:                        4799.99
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       192 KiB
L1i cache:                       192 KiB
L2 cache:                        24 MiB
L3 cache:                        16 MiB
NUMA node0 CPU(s):               0-5
Vulnerability Itlb multihit:     KVM: Mitigation: VMX unsupported
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Full generic retpoline, IBPB conditional, IBRS_FW, STIBP disabled, RSB fi
                                 lling
Vulnerability Srbds:             Unknown: Dependent on hypervisor status
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx f
                                 xsr sse sse2 ht syscall nx rdtscp lm constant_tsc rep_good nopl xtopology cpuid tsc_k
                                 nown_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_dea
                                 dline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm cpuid_fault invpcid_sing
                                 le pti ssbd ibrs ibpb fsgsbase bmi1 avx2 smep bmi2 erms invpcid xsaveopt arat

Running canary
real	1m56.190s
user	3m10.405s
sys	0m10.676s

Running with 1 concurrent jobs
real	1m56.727s
user	2m49.398s
sys	0m9.920s

Running with 2 concurrent jobs
real	1m12.947s
user	2m52.243s
sys	0m10.511s

Running with 3 concurrent jobs
real	0m59.123s
user	2m53.484s
sys	0m10.447s

Running with 4 concurrent jobs
real	0m54.922s
user	2m55.053s
sys	0m9.948s

Running with 5 concurrent jobs
real	0m52.627s
user	2m58.354s
sys	0m10.223s

Running with 6 concurrent jobs
real	0m49.624s
user	2m55.533s
sys	0m9.838s

Running with 7 concurrent jobs
real	0m48.919s
user	2m59.466s
sys	0m9.945s

Running with 8 concurrent jobs
real	0m47.643s
user	2m53.795s
sys	0m9.925s
```

File: cli/ops/test_dispatcher.rs
Comment (caspervonb) [reply cli/ops/test_dispatcher.rs]: Op has grown on me, I think it's fine.

File: cli/tests/integration_tests.rs
Comment (caspervonb) [reply cli/tests/integration_tests.rs:2431]: Punting this, needs additional work.

File: cli/main.rs
Comment (bartlomieju) [inline cli/main.rs:973]: Please move the body of this function to `cli/tools/test_runner.rs`

File: cli/ops/test_dispatcher.rs
Comment (bartlomieju) [inline cli/ops/test_dispatcher.rs:1]: Missing copyright

File: cli/test_dispatcher.rs
Comment (bartlomieju) [reply cli/test_dispatcher.rs:1]: This comment still stands

File: cli/tests/integration_tests.rs
Comment (bartlomieju) [reply cli/tests/integration_tests.rs:2431]: This should be reenabled before landing this PR, what's the problem here?

**Review by bartlomieju — CHANGES_REQUESTED** [review]

**Comment by caspervonb** [comment]
Discussed a bit with @bartlomieju, the diff on this patch is huge so to avoid having such a large diff in the tests, I'll be implementing permissions into Deno.test before merging this.

This patch is being punted to 1.10.

**Comment by caspervonb** [comment]
Blocked on #10188

File: cli/ops/test_runner.rs
Comment (caspervonb) [inline cli/ops/test_runner.rs:100]: When a permission isn't specified, should we leave it as is or drop it?

**Comment by caspervonb** [comment]
Blocked on #10330

File: cli/ops/test_dispatcher.rs
Comment (bartlomieju) [inline cli/ops/test_dispatcher.rs:3]: Nitpick: I think this module could be merged with `ops/test_runner.rs`

File: cli/tools/test_runner.rs
Comment (bartlomieju) [inline cli/tools/test_runner.rs:151]: `unload` handler is sync, no need to run event loop here

File: cli/tools/test_runner.rs
Comment (bartlomieju) [inline cli/tools/test_runner.rs:111]: I think the name of this function is confusing, does it run only a single test case from a file?

File: cli/tools/test_runner.rs
Comment (bartlomieju) [inline cli/tools/test_runner.rs:353]: Just to confirm - this output is matching the currently existing one, correct?

File: runtime/js/40_testing.js
Comment (bartlomieju) [inline runtime/js/40_testing.js:245]: 🚀  bye, bye `TestRunner`

File: cli/tools/test_runner.rs
Comment (caspervonb) [reply cli/tools/test_runner.rs:353]: Yup, to the tee including the ugly bits (raising that issue in a different PR).

Comment (caspervonb) [reply cli/tools/test_runner.rs:353]: An aside; this is ugly but have a branch where this whole block except the "control" bits (has_error, used_only) of the thread it are going into a `TestReporter`.

File: cli/ops/test_dispatcher.rs
Comment (caspervonb) [reply cli/ops/test_dispatcher.rs]: > Nitpick: I think this module could be merged with ops/test_runner.rs

Agreed, merged into ops::testing (js equivalent is called testing).

File: cli/tools/test_runner.rs
Comment (caspervonb) [reply cli/tools/test_runner.rs]: Renamed to run_test_file

File: cli/tests/integration_tests.rs
Comment (caspervonb) [reply cli/tests/integration_tests.rs]: Think we went over this in DMs last month or so but execute doesn't track unresolved promises.
So to handle this case we generate a tiny module that just calls `await runTests`.

File: cli/tools/test_runner.rs
Comment (bartlomieju) [reply cli/tools/test_runner.rs]: Thanks

**Comment by caspervonb** [comment]
**HOLD**

Looks like adding a module to call `runTests` caused a weird bug.

**Comment by bartlomieju** [comment]
I've noticed two quirks that I'm not sure how big of an impact they'll have:
1) There's `Check <module name>` for as many modules as there are discovered
```
Check file:///Users/biwanczuk/dev/deno_std/textproto/test.ts
Check file:///Users/biwanczuk/dev/deno_std/_util/deep_assign_test.ts
Check file:///Users/biwanczuk/dev/deno_std/_util/assert_test.ts
Check file:///Users/biwanczuk/dev/deno_std/flags/test.ts
Check file:///Users/biwanczuk/dev/deno_std/archive/tar_test.ts
Check file:///Users/biwanczuk/dev/deno_std/bytes/test.ts
Check file:///Users/biwanczuk/dev/deno_std/hash/_fnv/util_test.ts
Check file:///Users/biwanczuk/dev/deno_std/hash/sha512_test.ts
Check file:///Users/biwanczuk/dev/deno_std/hash/fnv_test.ts
Check file:///Users/biwanczuk/dev/deno_std/hash/sha256_test.ts
Check file:///Users/biwanczuk/dev/deno_std/hash/test.ts
Check file:///Users/biwanczuk/dev/deno_std/hash/sha1_test.ts
Check file:///Users/biwanczuk/dev/deno_std/hash/md5_test.ts
Check file:///Users/biwanczuk/dev/deno_std/hash/sha3_test.ts
Check file:///Users/biwanczuk/dev/deno_std/datetime/test.ts
Check file:///Users/biwanczuk/dev/deno_std/fmt/colors_test.ts
Check file:///Users/biwanczuk/dev/deno_std/fmt/printf_test.ts
....
```

2) For each test module (supposedly) there's `running X tests` printed:
```
running 4 tests
test [encoding/base64] testBase64EncodeString ... ok (2ms)
test [encoding/base64] testBase64EncodeBinary ... ok (2ms)
test [encoding/base64] testBase64EncodeBinaryBuffer ... ok (2ms)
test [encoding/base64] testBase64DecodeBinary ... ok (2ms)
running 4 tests
test [encoding.base32] encode ... ok (4ms)
test [encoding.base32] decode ... ok (2ms)
test [encoding.base32] decode bad length ... ok (1ms)
test [encoding.base32] decode bad padding ... ok (1ms)
```
I don't mind it, but maybe we could change it to `running X tests in <module name>`

**Comment by lucacasonato** [comment]
> There's `Check <module name> for as many modules as there are discovered`

This suggests each module is being type checked individually. AFAIK this is way slower than type-checking them all at once. cc @kitsonk

**Comment by caspervonb** [comment]
> 1. There's `Check <module name>` for as many modules as there are discovered
> 
> ```
> Check file:///Users/biwanczuk/dev/deno_std/textproto/test.ts
> Check file:///Users/biwanczuk/dev/deno_std/_util/deep_assign_test.ts
> Check file:///Users/biwanczuk/dev/deno_std/_util/assert_test.ts
> Check file:///Users/biwanczuk/dev/deno_std/flags/test.ts
> Check file:///Users/biwanczuk/dev/deno_std/archive/tar_test.ts
> Check file:///Users/biwanczuk/dev/deno_std/bytes/test.ts
> Check file:///Users/biwanczuk/dev/deno_std/hash/_fnv/util_test.ts
> Check file:///Users/biwanczuk/dev/deno_std/hash/sha512_test.ts
> Check file:///Users/biwanczuk/dev/deno_std/hash/fnv_test.ts
> Check file:///Users/biwanczuk/dev/deno_std/hash/sha256_test.ts
> Check file:///Users/biwanczuk/dev/deno_std/hash/test.ts
> Check file:///Users/biwanczuk/dev/deno_std/hash/sha1_test.ts
> Check file:///Users/biwanczuk/dev/deno_std/hash/md5_test.ts
> Check file:///Users/biwanczuk/dev/deno_std/hash/sha3_test.ts
> Check file:///Users/biwanczuk/dev/deno_std/datetime/test.ts
> Check file:///Users/biwanczuk/dev/deno_std/fmt/colors_test.ts
> Check file:///Users/biwanczuk/dev/deno_std/fmt/printf_test.ts
> ....
> ```

Doing the type checking and compilation before attempting to run is intentional.
Type errors occurs immediatly after the check statement so, it looks nice imo and accurately reflects what is going on.

> This suggests each module is being type checked individually. AFAIK this is way slower than type-checking them all at once. 

Yeah by one literally takes hours in the debug build @lucacasonato, but I do them all in one graph.

> 1. For each test module (supposedly) there's `running X tests` printed:
> 
> ```
> running 4 tests
> test [encoding/base64] testBase64EncodeString ... ok (2ms)
> test [encoding/base64] testBase64EncodeBinary ... ok (2ms)
> test [encoding/base64] testBase64EncodeBinaryBuffer ... ok (2ms)
> test [encoding/base64] testBase64DecodeBinary ... ok (2ms)
> running 4 tests
> test [encoding.base32] encode ... ok (4ms)
> test [encoding.base32] decode ... ok (2ms)
> test [encoding.base32] decode bad length ... ok (1ms)
> test [encoding.base32] decode bad padding ... ok (1ms)
> ```
> I don't mind it, but maybe we could change it to running X tests in <module name>

I've got a branch with *slightly* nicer test output to address that after this.

**Comment by caspervonb** [comment]
Ran into some caching issues after introducing the fake module stub that calls `await Deno[Deno.internal].runTests`.
For example switching from unit tests to std tests:

```
error: Module "file:///home/caspervonb/deno/test_util/std/fmt/printf_test.ts" is missing from the graph.
  From: file:///home/caspervonb/deno/$deno$test.ts
```

Also a few runs definitively used old cached versions of "$deno$test.ts" from disk.

File: cli/tools/test_runner.rs
Comment (bartlomieju) [inline cli/tools/test_runner.rs:251]: @piscisaureus please take a look at this part

**Review by bartlomieju — APPROVED** [review]
LGTM 🚀  thank you Casper, this is a great feature to have

**Pull request merged**
